zuul-core/src/main/java/com/netflix/zuul/netty/server/psk/TlsPskHandler.java

```diff
@@ -0,0 +1,499 @@
+        tlsPskServer =
+                new ZuulPskServer(new JcaTlsCryptoProvider().create(new SecureRandom()), registry, externalTlsPskProvider, ctx);
+        tlsPskServerProtocol = new TlsPskServerProtocol();
+        tlsPskServerProtocol.accept(tlsPskServer);
+        super.channelRegistered(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+    }
+
+    private static byte[] readDirect(ByteBuf byteBufMsg) {
+        int length = byteBufMsg.readableBytes();
+        byte[] dest = new byte[length];
+        byteBufMsg.readSlice(length).getBytes(0, dest);
+        return dest;
+    }
+
+    /**
+     * Returns the name of the current application-level protocol.
+     * Returns:
+     * the protocol name or null if application-level protocol has not been negotiated
+     */
+    public String getApplicationProtocol() {
+        return tlsPskServer!=null ? tlsPskServer.getApplicationProtocol() : null;
+    }
+
+    public SSLSession getSession() {
+        return tlsPskServerProtocol!=null ? tlsPskServerProtocol.getSSLSession() : null;
+    }
+
+    static class ZuulPskServer extends AbstractTlsServer {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(ZuulPskServer.class);
+
+        private static class PSKTimings {
+            private final Timer handshakeCompleteTimer;
+
+            private Long handshakeStartTime;
+
+            PSKTimings(Registry registry) {
+                handshakeCompleteTimer = registry.timer("zuul.psk.handshake.complete.time");
+            }
+
+            public void recordHandshakeStarting() {
+                handshakeStartTime = System.nanoTime();
+            }
+
+            public void recordHandshakeComplete() {
+                handshakeCompleteTimer.record(System.nanoTime() - handshakeStartTime, TimeUnit.NANOSECONDS);
+            }
+        }
+
+        private final PSKTimings pskTimings;
+
+        private final ExternalTlsPskProvider externalTlsPskProvider;
+
+        private final ChannelHandlerContext ctx;
+
+
+        public ZuulPskServer(
+                TlsCrypto crypto,
+                Registry registry,
+                ExternalTlsPskProvider externalTlsPskProvider, ChannelHandlerContext ctx) {
+            super(crypto);
+            this.pskTimings = new PSKTimings(registry);
+            this.externalTlsPskProvider = externalTlsPskProvider;
+            this.ctx = ctx;
+        }
+
+        @Override
+        public TlsCredentials getCredentials() {
+            return null;
+        }
+
+        @Override
+        protected Vector getProtocolNames() {
+            Vector protocolNames = new Vector();
+            protocolNames.addElement(ProtocolName.HTTP_1_1);
+            protocolNames.addElement(ProtocolName.HTTP_2_TLS);
```

Comment:
This list should probably be injected, esp. for clients we don't want to advertise h2 for.